src/core/worker_routine.hpp

```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+namespace docgen {
+namespace core {
+
+/*
+ * WorkerType must have type aliases for token_t and dest_t
+ */
+template <class WorkerType>
+struct WorkerRoutine
+{
+	using worker_t = WorkerType;
+	using tok_t = typename worker_t::token_t;
```

Comment:
`token_t`?

Reply:
Unfortunately, despite the fact that the `SomethingWorker_routines` struct inherits from this privately, it seems that the `SomethingWorker` struct which inherits from it still ultimately has access to these type aliases (this wouldn't be the case with typical members gawdammit), so we get an ambiguity conflict since `SomethingWorker` already inherits and uses `token_t` from `ParseWorker`.

Reply:
Wait no way, these aliases can be scoped by private also. If youcreate a class with a member alias in private, a user of this class canot directlyaccess that alias.

The reason why you cannot use token_t directly in this case is only because you have multiple inheritance and compiler assumes this token_t can come from either base class. I believe it gives this preemptive error whether token_t exists in one or both. A lot of ppl encourage not using inheritance even when you might find it intuitive because it becomes to hard to maintain. Once you inherit from something and later you have another class you want to inherit from, it's multiple inheritance and it's so easy to be in this situation. I've thought about this a lot and here's my advice:

- if it makes sense to think of these base classes rather as policies (template parameters of the class) do that
- if you're deriving from a static class, no need to inherit.
- in either case, if you have a static class, simply access its members through the class name::something inside your current class like how I commented on the last update.
- if it's a non-static class, store an object of that type as a member rather than inheriting. (Inheritance ultimately has to do this underneath when compiler compiles) 
- if you must must insist on multiple inheritance, re-expose the types in current class like

using token_t = worker_class_t::token_t;

This is again not considered code repetition. It's just like delegating function calls. This is saying I have an interface for this class which delegates implementation to worker_class_t.